src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,444 @@
+        }
+    }
+}
+
+void GUPSGen::wakeUp()
+{
+    if (!nextSendEvent.scheduled() && !requestPool.empty())
+    {
+        schedule(nextSendEvent, nextCycle());
+    }
+}
+
+void
+GUPSGen::createNextReq() {
+
+    if (!responsePool.empty()) {
+        PacketPtr pkt = responsePool.front();
+        responsePool.pop();
+
+        uint64_t *updated_value = pkt->getPtr<uint64_t>();
+        *updated_value ^= updateTable[pkt->req];
+        updateTable.erase(pkt->req);
+
+        uint8_t write_data[elementSize];
+        std::memcpy(write_data, updated_value, elementSize);
+        Addr addr = pkt->getAddr();
+        delete pkt;
+
+        PacketPtr new_pkt = getWritePacket(addr, elementSize, write_data);
+        requestPool.push(new_pkt);
+    }
+    else if (!doneReading) {
+        assert (rIndex < 128);
+
+        randomSeeds[rIndex] = (randomSeeds[rIndex] << 1) ^
+                        ((int64_t) randomSeeds[rIndex] < 0 ? POLY : 0);
+        uint64_t value = randomSeeds[rIndex];
+        uint64_t index = randomSeeds[rIndex] & (tableSize - 1);
+        Addr addr = indexToAddr(index);
+        PacketPtr pkt = getReadPacket(addr, sizeof(uint64_t));
+        updateTable[pkt->req] = value;
+        requestPool.push(pkt);
+
+        rIndex++;
+        if (rIndex >= 128) {
+            rIndex = 0;
+            iterations++;
+            if (iterations >= numUpdates / 128) {
+                DPRINTF(GUPSGen, "sendNextReq: Done creating reads.\n");
+                doneReading = true;
+            }
+        }
+    }
+
+    if (!nextCreateEvent.scheduled()) {
+        if ((requestPool.size() < reqQueueSize) &&
+            (!doneReading || !responsePool.empty())) {
+            schedule(nextCreateEvent, nextCycle());
+        }
+    }
+
+    if (!nextSendEvent.scheduled()) {
+        if (!requestPool.empty()) {
+            schedule(nextSendEvent, nextCycle());
+        }
+    }
+
+    return;
+}
+
+void
+GUPSGen::sendNextReq() {
+
+    if (!port.blocked())
+    {
+        PacketPtr pkt = requestPool.front();
+
+        exitTimes[pkt->req] = curTick();
+        if (pkt->isWrite()) {
+            DPRINTF(GUPSGen, "sendNextReq: Sent write pkt, pkt->addr_range: "
+                            "%s, pkt->data: %lu.\n",
+                            *pkt->getPtr<uint64_t>(),
+                            pkt->getAddrRange().to_string());
+        }
+        else {
+            DPRINTF(GUPSGen, "sendNextReq: Sent read pkt, pkt->addr_range: "
+                            "%s.\n", pkt->getAddrRange().to_string());
+        }
+        port.sendTimingPacket(pkt);
+        onTheFlyRequests++;
+        DPRINTF(GUPSGen, "%s: onTheFlyRequests: %d.\n",
+                __func__, onTheFlyRequests);
+        requestPool.pop();
+    }
+
+    if (!nextCreateEvent.scheduled()) {
+        if ((requestPool.size() < reqQueueSize) &&
+            (!doneReading || !responsePool.empty())) {
+            schedule(nextCreateEvent, nextCycle());
+        }
+    }
```

Comment:
Do you do this even if the port is blocked?

Reply:
I do this since the functionality of the createNextReq function does not interfere with the port. It will stop creating the requests only if the requestPool is full (has reached reqQueueSize). The port applies back pressure to the creator through the requestPool.